deployment/postgres-init-db/sql/tables/merlin/external_events/plan_derivation_group.sql

```diff
@@ -0,0 +1,27 @@
+create table merlin.plan_derivation_group (
```

Comment:
Just had a chat with @duranb about this table, and he asked if it could be tweaked such that the UI is only setting a boolean `acknowledged` column (or sth named similarly), and the DB is setting an `updated_at`/`last_acknowledged_at` field via an update trigger, as the general pattern is that the UI does not set these timestamps itself.

The answer is yes, this is possible, but would require two triggers: one on `AFTER INSERT` of an external source that would set `acknowledged` to `false` and one `BEFORE UPDATE` on this table that only sets `last_acknowledged_at` to `now()` if and only if `acknowledged` is being updated to `true`.

Please let me know your thoughts on this proposal.

Reply:
Chatted some more with @duranb about this proposal and as another argument in favor of this, it would allow the UI to only select rows where `acknowledged` is `false` (as in, there are unack-ed changes), instead of needing to select all rows from the plan and doing comparisons to discover if there any changes.

Reply:
To me, this seems like a fair amount of work and more importantly, complexity, being introduced for what seems to be a non-issue in the UI. The code savings/simplifications in the UI really do seem minimal, as this is already going to be a pretty small table to start. There are three parts of the UI that would change:

**In `ExternalSourcesPanel.svelte`**:
Currently, our filtering mechanism to determine the set of sources that are yet unacknowledged in our update card is as follows:
```
// Determine which new and deleted sources are unacknowledged for the user
  $: {
    if ($plan !== null && $plan.id !== null && $derivationGroupsLastAcknowledged[$plan.id] !== undefined) {
      // a mapping of derivation groups (that are associated with this plan) to the time their last update was last acknowledged
      const planDerivationGroupLastAcknowledgedAt = $derivationGroupsLastAcknowledged[$plan.id];
      const groups = Object.keys(planDerivationGroupLastAcknowledgedAt);
      unseenSources = $externalSources.filter(externalSource => {
        if (!groups.includes(externalSource.derivation_group_name)) {
          return false;
        }
        return (
          new Date(externalSource.created_at) >
          new Date(planDerivationGroupLastAcknowledgedAt[externalSource.derivation_group_name])
        );
      });
    }
  }
```
This change modifies that last line, from:
```
return (
  new Date(externalSource.created_at) >
  new Date(planDerivationGroupLastAcknowledgedAt[externalSource.derivation_group_name])
);
```
to
```
return planDerivationGroupLastAcknowledgedAt[externalSource.derivation_group_name];
```

There is also a function that changes slightly, as an effect that normally passes a date will no longer pass any value (as updates from the UI are equivalent to setting `acknowledged` to true; the only thing that could set it false is an `external_source` update):
```
function onUpdateDismiss() {
  for (const derivationGroup of filteredDerivationGroups) {
    effects.updateDerivationGroupAcknowledged($plan ?? undefined, derivationGroup.name, new Date(), user);
  }
}
```

```
function onUpdateDismiss() {
  for (const derivationGroup of filteredDerivationGroups) {
    effects.updateDerivationGroupAcknowledged($plan ?? undefined, derivationGroup.name, user);
  }
}
```

In `effects.ts`:
Now, the effect `` will not take a date, but rather pass the value `true` in its effect, as described earlier.

```
async updateDerivationGroupAcknowledged(
  plan: Plan | undefined,
  derivation_group_name: string,
  updatedAt = new Date(),
  user: User | null,
) {
  if (plan === undefined) {
    return;
  }
  try {
    if (!queryPermissions.UPDATE_DERIVATION_GROUP_ACKNOWLEDGED(user, plan)) {
      throwPermissionError('mark viewership of a updates to a derivation group');
    }
    const { updatePlanDerivationGroup: update } = await reqHasura(
      gql.UPDATE_DERIVATION_GROUP_ACKNOWLEDGED,
      { derivation_group_name, new_date: updatedAt.toUTCString(), plan_id: plan.id },
      user,
    );
    if (update) {
      return update;
    } else {
      throw Error(`Unable to log derivation group update recognition`);
    }
  } catch (e) {
    catchError('Derivation Group Update Visibility Recognition Failed', e as Error);
  }
},
```
becomes
```
async updateDerivationGroupAcknowledged(
  plan: Plan | undefined,
  derivation_group_name: string,
  user: User | null,
) {
  if (plan === undefined) {
    return;
  }
  try {
    if (!queryPermissions.UPDATE_DERIVATION_GROUP_ACKNOWLEDGED(user, plan)) {
      throwPermissionError('mark viewership of a updates to a derivation group');
    }
    const { updatePlanDerivationGroup: update } = await reqHasura(
      gql.UPDATE_DERIVATION_GROUP_ACKNOWLEDGED,
      { derivation_group_name, acknowledged: true, plan_id: plan.id },
      user,
    );
    if (update) {
      return update;
    } else {
      throw Error(`Unable to log derivation group update recognition`);
    }
  } catch (e) {
    catchError('Derivation Group Update Visibility Recognition Failed', e as Error);
  }
},
```

I'm not sure if this warrants an update, and the introduced complexity with a trigger on `external_source` seems like unnecessary complexity.

Reply:
My concern was that these types of timestamps aren't conventionally user provided.  As an API, we open it up to the user potentially providing the wrong timestamp for a feature that's really just "did this get dismissed by someone". It's a small distinction and I don't know the best practices for DB so I'll defer to @Mythicaeda and @mattdailis.

The benefit to the UI is that it could leverage the boolean to determine whether or not it should iterate through all (or a smaller subset of) the groups in the first place. Without it, it will go through all the groups regardless everytime the page is loaded just to show a message that shows up potentially once.

Reply:
1) I agree with Bryan on the DB conventions and said as much in my initial comment
2) There isn't too much to iterate on _so far in your testing_. That is _not_ a good reason to not make the UI have to request and process less data.

Reply:
Very well; it has been updated in [add `acknowledged` boolean field to plan_derivation_group.sql](https://github.com/NASA-AMMOS/aerie/commit/d269e788b0dad764c5b983da7f65c733baf20d68). The change will be reflected on the frontend shortly.